### PR TITLE
SDI-477 Implement error handler to immediately push failed messages onto the DLQ

### PR DESCRIPTION
**Pull request by mikehalmamoj**
For some reason Spring Cloud AWS waits for the message visibility timeout to retry failed messages whereas the old library we used (AWS Java messaging lib) [sets the visibility timeout to zero](https://github.com/awslabs/amazon-sqs-java-messaging-lib/blob/8627d032d5e8526774b22a93db978bfed98b5e96/src/main/java/com/amazon/sqs/javamessaging/acknowledge/NegativeAcknowledger.java) when a message fails hence forcing the failed message to be retried immediately.

This PR implements the same solution by adding an error handler for failed messages that adjusts the visibility timeout to zero so that failed messages are retried immediately. We've also added a new config property `errorVisibilityTimeout`  so that this value can be extended if a longer interval between retries is required.

This PR contains a few other random changes to tidy things up:
* move to latest Localstack
* align the `test-app` and `test-app-async` so the only differences are that one uses Spring Webflux and the other Spring WebMVC
* replace incorrect usage of GlobalScope (which we've since learned is a bad thing and was actually breaking some tests now that DLQ processing works correctly)

### CODE REVIEW

File: hmpps-sqs-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueFactory.kt [hunk 1]
```diff
@@ -60,15 +62,23 @@ class HmppsQueueFactory(
...
       .builder<Any>()
       .sqsAsyncClient(awsSqsClient)
+      .errorHandler(object : ErrorHandler<Any> {
+        override fun handle(message: org.springframework.messaging.Message<Any>, t: Throwable) {
+          // SDI-477 remove this logging when we are comfortable that all is working as expected - instant retries
+          log.info("Setting visibility of messageId ${message.headers["id"]} to $errorVisibilityTimeout (to initiate faster retry) after receiving exception ${t.cause?.cause?.cause}")
+          val sqsVisibility = message.headers["Sqs_Visibility"] as QueueMessageVisibility
+          sqsVisibility.changeTo(errorVisibilityTimeout)
+        }
+      })
```
Comment (mikehalmamoj) [inline hmpps-sqs-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsQueueFactory.kt:81]: The main change to timeout messages in error immediately (errorVisibilityTimeout defaults to 0).

File: test-app-reactive/src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstemplatepackagenameasync/integration/HmppsEventProcessingTest.kt [hunk 2]
```diff
@@ -63,14 +64,32 @@ class HmppsEventProcessingTest : IntegrationTestBase() {
...
+      .let { objectMapper.readValue(it, Message::class.java) }
+    val receivedEvent = objectMapper.readValue(message, HmppsEvent::class.java)
 
     assertThat(receivedEvent.id).isEqualTo("event-id")
     assertThat(receivedEvent.type).isEqualTo("offender.movement.reception")
     assertThat(receivedEvent.contents).isEqualTo("some event contents")
   }
+
+  @Test
+  fun `event is moved to the dead letter queue when an exception is thrown`() = runBlocking<Unit> {
```
Comment (mikehalmamoj) [inline test-app-reactive/src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstemplatepackagenameasync/integration/HmppsEventProcessingTest.kt:78]: Test was in `testapp` but not `test-app-reactive`

File: test-app/src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstemplatepackagename/service/MessageListener.kt [hunk 3]
```diff
@@ -2,15 +2,21 @@ package uk.gov.justice.digital.hmpps.hmppstemplatepackagename.service
...
+import kotlinx.coroutines.future.future
 import org.springframework.stereotype.Service
+import java.util.concurrent.CompletableFuture
 
 @Service
 class InboundMessageListener(private val inboundMessageService: InboundMessageService, private val objectMapper: ObjectMapper) {
 
   @SqsListener("inboundqueue", factory = "hmppsQueueContainerFactoryProxy")
-  fun processMessage(message: Message) {
+  fun processMessage(message: Message): CompletableFuture<Void> {
```
Comment (mikehalmamoj) [inline test-app/src/main/kotlin/uk/gov/justice/digital/hmpps/hmppstemplatepackagename/service/MessageListener.kt:15]: Both `test-app` and `test-app-reactive` now have 1 synchronous and 1 asynchronous message listener

File: test-app/src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstemplatepackagename/integration/HmppsEventProcessingTest.kt [hunk 4]
```diff
@@ -86,39 +87,9 @@ class HmppsEventProcessingTest : IntegrationTestBase() {
...
         .build()
-    ).get()
+    )
 
     await untilCallTo { inboundSqsDlqClient.countMessagesOnQueue(inboundDlqUrl).get() } matches { it == 1 }
     assertThat(inboundSqsClient.countAllMessagesOnQueue(inboundQueueUrl).get()).isEqualTo(0)
   }
-
-  @Test
-  fun `event is not visible on queue when an exception is thrown`() {
```
Comment (mikehalmamoj) [inline test-app/src/test/kotlin/uk/gov/justice/digital/hmpps/hmppstemplatepackagename/integration/HmppsEventProcessingTest.kt:96]: Remove this test which was demonstrating previous incorrect behaviour.

**Review by andymarke — APPROVED** [review]
lgtm

**Pull request merged**